### ISSUE TITLE
3. Linear Regression 에서의 cost 최소화 알고리즘 원리

### ISSUE THREAD

**Issue by NahahaNahaha** [NahahaNahaha/tensorflow#4]
기존에 배웠던 H(x)= Wx+ b에서 
cost 함수는  
<img width="203" alt="2018-03-30 1 16 25" src="https://user-images.githubusercontent.com/26895961/38124466-c4979470-341c-11e8-8fe8-ed9b25c4e667.png"> 
과 같다. 하지만 cost함수의 최소화를 좀 간단하게 보기 위해서 H(x)= Wx 로 간단화 시킨다. 
<img width="686" alt="2018-03-30 1 21 09" src="https://user-images.githubusercontent.com/26895961/38124539-40ff35ea-341d-11e8-8994-7e16a49c8a5a.png">
W의 값에 변화 따른 cost 의 변화값을 함수로 나타내면 이렇게 표현되는데 (training set 이 x=[1,2,3], y=[1,2,3]인 경우, 즉 가장 이상적인 W의 값이 1인 경우) 
W=1일 때를 축으로 가지는 이차함수 형태인 것을 볼 수 있다. 
이제 여기서 이상적인 W의 값은 Gradient descent algorithm를 이용해서 찾을 수 있다. 
우선 아무 값이 W를 설정하고 여기서 W와 b를 아주 미세하게 변화 시킨다. 그래서 cost(W,b)의 값이 작아진다면 그 방향(gradient)로 지속적으로 조금씩 변화시킨다. 이를 계속 반복하면 이상적인 W의 값을 찾을 수 있다. 
수식으로 나타내면 아래와 같이 나타낼 수 있다.  
우선 미분을 한 뒤 비교적 깔끔하게 나타내기 위해 1/m 이 아닌 1/2m으로 설정한다. 
<img width="598" alt="2018-03-30 1 28 43" src="https://user-images.githubusercontent.com/26895961/38124668-63c64432-341e-11e8-83b1-1975602bbb12.png">
그 후 상수 알파를 이용해서 위처럼 나타낼 수 있다. 왜냐하면 지금의 케이스같은 경우 cost 함수의 기울기가 양수인 경우 W의 값을 줄여야 cost의 값이 줄어듦으로 기존의 W에서 미세한 값을 빼어야하고 반대로 기울기가 음수인 경우는 기존의 W에서 미세한 값을 더해주어야 함으로 수식이 저렇게 표현된다. (여기서 알파값이 중요한 것 같은데 알파가 작으면 작을수록 좀 더 정밀하게 이상적인 W값을 찾을 수 있지만 연산을 많이 해야하고 크면 클수록 정밀하게 찾을 수는 없지만 연산을 덜 해도 되니 적당한 값을 찾는게 중요한 것같다. )
식을 풀면 아래와 같다.
<img width="746" alt="2018-03-30 1 28 05" src="https://user-images.githubusercontent.com/26895961/38124640-34fe12ba-341e-11e8-9e84-32c2d17d2637.png">
따라서  Gradient descent algorithm를 수식으로 나타내어 보았다,. 
여기 나온 케이스는 2차원적으로 이차함수 형태이고 3차원으로 나타내어도 다음과 같이 표현되기 때문에 
<img width="1006" alt="2018-03-30 1 35 19" src="https://user-images.githubusercontent.com/26895961/38124758-3d7db494-341f-11e8-979a-ac3eea374aa2.png">
모든 경사면에서 cost의 최소값으로 가는 루트가 정해져있다. (최소값이 한 개이므로)  
하지만 W와 b 그리고 cost의 3차원 함수가 다음과 같이 생성되면  
<img width="675" alt="2018-03-30 1 37 13" src="https://user-images.githubusercontent.com/26895961/38124787-7c0145be-341f-11e8-8a5f-d83c48820194.png">
최소로 가는 루트가 여러가지이므로 잘못된 루트로 진행될 수 있다. 이게 단점인듯 한데 어떻게 보완할지 궁금하다.
